src/realm/object-store/sync/sync_session.cpp

```diff
@@ -888,11 +888,18 @@ static sync::Session::Config::ClientReset make_client_reset_config(RealmConfig s
                                                           did_recover);
         };
     }
-    if (session_config.sync_config->notify_before_client_reset) {
-        config.notify_before_client_reset = [config = session_config](VersionID version) {
-            config.sync_config->notify_before_client_reset(Realm::get_frozen_realm(config, version));
-        };
-    }
+    bool notify_user_before = bool(session_config.sync_config->notify_before_client_reset);
+    config.notify_before_client_reset = [notify_user_before, config = session_config]() -> VersionID {
+        auto frozen = Realm::get_shared_realm(config)->freeze();
```

Comment:
do we need something similar for the frozen realm in `notify_after_client_reset`?

Reply:
It isn't necessary for the _after_ callback under the assumption that the schema cannot be changed on the fly since the _before_ callback was invoked. If the user supplies an _after_ callback, then at the time of the reset, the _before_ state is frozen and pinned so that that frozen version is still available at the time of the _after_ callback.